src/lib.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2017 King's College London
+// created by the Software Development Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any person obtaining a
+// copy of this software, associated documentation and/or data (collectively the "Software"), free
+// of charge and under any and all copyright rights in the Software, and any and all patent rights
+// owned or freely licensable by each licensor hereunder covering either (i) the unmodified
+// Software as contributed to or provided by such licensor, or (ii) the Larger Works (as defined
+// below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software is contributed
+// by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create derivative works
+// of, display, perform, and distribute the Software and make, use, sell, offer for sale, import,
+// export, have made, and have sold the Software and the Larger Work(s), and to sublicense the
+// foregoing rights on either these or other terms.
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+
+const HOT_THRESHOLD: u16 = 10;
+
+/// What was the outcome of a `jit_control_point`?
+#[derive(Debug, PartialEq)]
+pub enum ControlOutcome {
+    /// The `jit_control_point` executed a trace: the interpreter does not need to execute a
+    /// trace for this location in the end user's code.
+    ExecutedTrace,
+    /// The `jit_control_point` did not execute a trace: the interpreter needs to execute the
+    /// relevant code for this location in the end user's code.
+    DidntExecuteTrace
+}
+
+/// The current meta-tracing state of a given location in the end-user's code.
+#[derive(Debug)]
+enum LocCount {
+    /// Below `HOT_THRESHOLD`, but counting up towards it.
+    Counting(u16),
```

Comment:
Shall we alias a type for the counter, like before?

Reply:
In general I'm not a big fan of type aliases because they don't imply any type constraints. [Two types `T1` and `T2` both aliased to `u16` are considered entirely equivalent in Rust.] That said, if we need to hide the details of a type, aliasing can be an easy hack. Since this is an internal detail, I'm not sure it's worth the extra effort.

Reply:
I see. Well, if there is a way to have a type-distinct counter type, I'd be all for it.

Reply:
There are ways, but they're heavyweight. They're only worth doing if you're worried that you will accidentally coerce between two types you would like to be distinct. That isn't a real worry here.

Reply:
Fine, let's leave it for now.